torch/special/__init__.py

```diff
@@ -799,3 +794,65 @@
     tensor([1., 1., 1.])
 
 """.format(**common_args))
+
+
+def _logsumexp_return_sign_false(
+        input: Tensor, dim: typing.List[int], keepdim: bool = False,
+        b: typing.Optional[Tensor] = None, return_sign: bool = False, out: Tensor = None) -> Tensor:
+    return _special.special_logsumexp(input, dim, keepdim, b, out=out)
+
+
+def _logsumexp_return_sign_true(
+        input: Tensor, dim: typing.List[int], keepdim: bool = False,
+        b: typing.Optional[Tensor] = None, return_sign: bool = False,
+        out: typing.Tuple[Tensor, Tensor] = None) -> typing.Tuple[Tensor, Tensor]:
+    r"""
+    Returns the log of summed exponentials of each row of the :attr:`input`
+    tensor (optionally scaled by :attr:`b`) in the given dimension :attr:`dim`.
+    The computation is numerically stabilized.
+
+    For summation index :math:`j` given by `dim` and other indices :math:`i`, the result is
+
+        .. math::
+            \text{logsumexp}(x)_{i} = \log \sum_j b_{ij} \exp(x_{ij})
+
+
+    If :attr:`keepdim` is ``True``, the output tensor is of the same size
+    as :attr:`input` except in the dimension(s) :attr:`dim` where it is of size 1.
+    Otherwise, :attr:`dim` is squeezed (see :func:`torch.squeeze`), resulting in the
+    output tensor having 1 (or ``len(dim)``) fewer dimension(s).
+
+
+    Args:
+        input (Tensor): the input tensor.
+        dim (int or tuple of ints): the dimension or dimensions to reduce.
+        keepdim (bool): whether the output tensor has :attr:`dim` retained or not.
+        b (Tensor): Scaling factors for exponentiated :attr:`input` (may be negative).
+        return_sign (bool): whether to return sign information (negative results will
+            return NaN if :attr:`return_sign` is false).
+
+    Keyword args:
+        out (tuple of Tensors, optional): output tensor for the result and sign information. The
+            second element of the tuple will remain unchanged if :attr:`return_sign` is false.
+
+    Example::
+
+        >>> a = torch.randn(3, 3)
+        >>> torch.logsumexp(a, 1)
+        tensor([1.4907, 1.0593, 1.5696])
+        >>> torch.dist(torch.logsumexp(a, 1), torch.log(torch.sum(torch.exp(a), 1)))
+        tensor(1.6859e-07)
+    """
+    return _special._special_logsumexp_with_sign(input, dim, keepdim, b, return_sign, out=out)
+
+
+logsumexp = boolean_dispatch(
```

Comment:
Instead of supporting torchscript what about creating a Python wrapper and choosing which function to call based on the value of `return_sign`? If `return_sign` is True and the user is scripting an error can be thrown.

Reply:
I played around with that a bit, but the problem is that the compiler needs to know which branch it will take at compile time rather than at runtime (or I couldn't just couldn't figure out the right signature to have the compiler accept the python wrapper).